File: 프로그래머스/레벨 2/프로그래머스  수식 최대화.cpp


#include <string>
#include <vector>
#include <algorithm>

using namespace std;
#define ll long long

ll cal(ll a, ll b, char op){
    if(op == '+')
        return a + b;
    else if(op == '-')
        return a - b;
    return a * b;
}
// 계산해주는 함수

long long solution(string expression) {
    long long answer = 0;
    
    string st = "*+-";
    vector<ll>v;
    string tmp;
    
    string operand;
    
    sort(st.begin(), st.end());
    
    for(char k : expression){
        if(k >= '0' && k <= '9')
            tmp += k;
        // k가 숫자일 경우 tmp에 더해다 줌
        // 여기서 숫자로 안하고 string으로 더하므로 10곱하고 그런거 안해도댐
        else{
            v.push_back(stoi(tmp));
            operand += k;
            tmp = "";
        }
        // 부호가 나올 경우 v에 이제까지 있던 tmp를 int로 바꾸어 넣어준다.
        // 그러고 부호를 operand에 저장해놓고, tmp도 초기화 시켜줌
    }
    v.push_back(stoi(tmp));
    // 남은 것도 넣어다 줌
    
    do{
        string str = operand;
        // 이제까지 들어간 부호들을 str에 넣어준다.
        vector<ll>v2 = v;
        
        for(char k : st){
            for(int i=0;i<str.size();i++){
                // 들어간 부호들을 조사
                if(k == str[i]){
                    // st를 돌리면서 str과 부호가 같은 경우
                    ll cnt = cal(v2[i], v2[i+1], k);
                    // v에는 각각 숫자들이 들어감으로 그거에 관해 계산한다.
                    
                    v2.erase(v2.begin() + i);
                    v2.erase(v2.begin() + i);
                    // v2에서 쓴 2가지 수를 지우고
                    
                    v2.insert(v2.begin() + i, cnt);
                    str.erase(str.begin() + i);
                    i--;
                    // v2에 계산한 cnt를 다시 넣어준다.
                    // 그러고 사용한 str 즉 부호를 지워주고,
                    // i는 하나만 줄여줌 ( 해야한다 )
                }
            }
        }
        answer = max(answer, abs(v2[0]));
        // v2[0] 즉 계산 된 값이 answer의 최대 보다 클 경우 갱신해준다.
        
    }while(next_permutation(st.begin(), st.end()));
    
    
    return answer;
}

File: 프로그래머스/레벨 2/프로그래머스 수식 최대화.py
from itertools import permutations
from collections import deque

def cal(priority, expression):
    arr = deque()
    st = ''
    
    for k in expression:
        if k.isdigit():
            st += k
            # k가 숫자일 경우 숫자를 문자 그대로 st에 더해준다.
        else:
            arr.append(st)
            st = ''
            arr.append(k)
            # k가 기호일 경우 들어온 st를 arr에 넣고 초기화한후
            # 기호도 arr에 넣어준다.
    arr.append(st)
    # 마지막의 st가 안들어오므로 arr에 넣어줌
    
    for op in priority:   # 기호의 순서대로인데 처음의 op를 먼저
        stack = []
        while len(arr) != 0:   # arr이 없어질때까지
            tmp = arr.popleft()
            if tmp == op:   # 만약 tmp와 op가 같을 경우에 계산을 해준다.
                result = str(eval(stack.pop() + op + arr.popleft()))
                stack.append(result)
                # stack의 직전 값과 op, arr를 한번 더 팝하여 다음 값을 구한다.
                # result 도 stack에 넣어줌
            else:
                stack.append(tmp)
                # op가 아닐 경우 stack에 다 넣어준다.
        arr = deque(stack)
        # 없어 진 경우 stack을 arr에 다시 넣고 다음 기호에 대해 진행해줌
        
    return int(arr.pop())
    # 다 끝나고 마지막 남은 arr를 pop해주면서 값을 리턴한다.
    
    

def solution(expression):
    answer = 0
    
    # + - * 의 순위를 마음대로
    for priority in list(permutations(['+', '-', '*'], 3)):
        answer = max(answer, abs(cal(priority, expression)))
    # 순서가 바뀔때마다 진행해줌
    
    return answer
